lib/buffer-based/writable.js

```diff
@@ -0,0 +1,412 @@
+    encoding = null;
+  }
+
+  if (isBuffer) {
+    encoding = 'buffer';
+  } else if (!encoding) {
+    encoding = state.defaultEncoding;
+  }
+
+  if (typeof callback !== 'function') {
+    callback = function() {};
+  }
+
+  if (state.ending) {
+    writeAfterEnd(this, callback);
+    return false;
+  }
+
+  if (isBuffer || isChunkValid(this, state, chunk, callback)) {
+    state.pendingcb++;
+    return writeOrBuffer(this, state, isBuffer, chunk, encoding, callback);
+  }
+
+  return false;
+};
+
+Writable.prototype.end = function(chunk, encoding, callback) {
+  const state = this._writableState;
+
+  if (typeof chunk === 'function') {
+    callback = chunk;
+    chunk = null;
+    encoding = null;
+  } else if (typeof encoding === 'function') {
+    callback = encoding;
+    encoding = null;
+  }
+
+  if (chunk !== null && chunk !== undefined)
+    this.write(chunk, encoding);
+
+  if (state.corked) {
+    state.corked = 1;
+    this.uncork();
+  }
+
+  if (!state.ending) {
+    endWritable(this, state, callback);
+  }
+
+  return this;
+};
+
+Writable.prototype.cork = function() {
+  this._writableState.corked++;
+};
+
+Writable.prototype.uncork = function() {
+  const state = this._writableState;
+
+  if (state.corked) {
+    state.corked--;
+
+    if (
+      !state.writing &&
+      !state.corked &&
+      !state.bufferProcessing &&
+      state.bufferOffset
+    ) {
+      clearBuffer(this, state);
+    }
+  }
+};
+
+Writable.prototype.pipe = function() {
+  this.emit('error', new Error('Cannot pipe, not readable'));
+};
+
+// internal functions
+
+const writeAfterEnd = (stream, callback) => {
+  const error = new Error('write after end');
+
+  stream.emit('error', error);
+  process.nextTick(callback, error);
+};
+
+function isChunkValid(stream, state, chunk, callback) {
+  let error;
+
+  if (chunk === null) {
+    error = new Error('May not write null values to stream');
+  } else if (typeof chunk !== 'string' && !state.objectMode) {
+    error = new Error('Chunk must be an instance of Buffer or string');
+  } else if (
+    state.objectMode &&
+    (typeof chunk.valueOf !== 'function' || (chunk.valueOf && chunk.valueOf() === chunk)) &&
+    typeof chunk[Symbol.toPrimitive] !== 'function'
+  ) {
+    error = new Error('Object has to have \'valueOf()\' or \'[Symbol.toPrimitive]()\' methods');
+  }
+  if (error) {
+    stream.emit('error', error);
+    process.nextTick(callback, error);
+    return false;
+  }
+
+  return true;
+};
+
+const writeOrBuffer = (stream, state, isBuffer, chunk, encoding, callback) => {
+  if (!isBuffer) {
+    const newChunk = decodeChunk(state, chunk, encoding);
+
+    if (chunk !== newChunk) {
+      isBuffer = true;
+      encoding = 'buffer';
+      chunk = newChunk;
+    }
+  }
+
+  const length = state.objectMode ? 1 : chunk.length;
+  state.length += length;
+
+  let ret =
+    state.length < state.highWaterMark &&
+    state.bufferOffset + Buffer.from(chunk).length < state.internalBuffer.length;
```

Comment:
- make separate variable for Buffer.from(chunk)
- use state.highWaterMark instead of state.internalBuffer.length because they have the same value and in the 1st case you have less nesting